mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala

```diff
@@ -94,15 +99,19 @@ class HashingTF @Since("1.4.0") (@Since("1.4.0") override val uid: String)
   @Since("2.0.0")
   override def transform(dataset: Dataset[_]): DataFrame = {
     val outputSchema = transformSchema(dataset.schema)
-
-    val hashingTF = new OldHashingTF($(numFeatures)).setBinary($(binary))
-    val func = (terms: Seq[_]) => {
-      val seq = hashingTF.transformImpl(terms)
-      Vectors.sparse(hashingTF.numFeatures, seq)
+    val hashUDF = udf { (terms: Seq[_]) => {
+      val termFrequencies = mutable.HashMap.empty[Int, Double]
+      val setTF =
+        if ($(binary)) (i: Int) => 1.0 else (i: Int) => termFrequencies.getOrElse(i, 0.0) + 1.0
+        terms.foreach { term =>
+          val i = Utils.nonNegativeMod(hashFunc(term), $(numFeatures))
+          termFrequencies.put(i, setTF(i))
```

Comment:
I may be overthinking this, but it might be faster to not get the value for the key and then update the value for the key. If the mutable map has `.withDefaultValue(0.0)` then...
```
if ($(binary)) {
  termFrequences(i) = 1.0
} else {
  termFrequences(i) += 1.0
}
```
It may mean reading `$(binary)` to a local val to make sure it's not accessed repeatedly. Or just duplicate the foreach over terms in two cases so that binary is checked only once.